Handle requests with headers

MaggieVL opened the pull request.

Commit:
Add a request header to the api

So that I can use it to test ServantElm.

Commit:
Add server-side header generation

Commit:
Add generated queries with headers

Commit:
Add client-side query invocation

googleson78 commented on examples/backend/app/Main.hs:
```diff
@@ -8,16 +8,17 @@ import Elm.Generate (Settings, defaultSettings, generateElm)
 main = do
   generateElm @Types settings
   generateElmModule settings userAPI
+  --print $ getReqs userAPI
```
Commented out code, delete this.

MaggieVL replied:
done

googleson78 commented on examples/backend/app/Main.hs:
```diff
@@ -8,16 +8,17 @@ import Elm.Generate (Settings, defaultSettings, generateElm)
 import Network.Wai.Handler.Warp (run)
 import Network.Wai.Middleware.Cors (CorsResourcePolicy (..), cors, simpleCorsResourcePolicy)
-import ServantElm (generateElmModule)
+import ServantElm (generateElmModule {-, getReqs-})
```
Commented out code, delete this.

MaggieVL replied:
done

googleson78 commented on examples/backend/src/ExampleAPI.hs:
```diff
@@ -19,18 +19,19 @@ module ExampleAPI
   "users" :> Get '[JSON] [User]
     :<|> "albert" :> Get '[JSON] User
     :<|> "signup" :> ReqBody '[JSON] User :> Post '[JSON] User
-    :<|> "sth" :> Post '[JSON] User
+    :<|> "sth" :> Header "someHeader" Text :> Post '[JSON] Text
```
I think you should add another endpoint with a non-trivial to encode value (e.g. not `Int` or `Text`), so it shows how we decided deal with those for now.

I think we should also document things at this point, as this decision (and the reasoning behind it) isn't obvious.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -250,6 +272,23 @@ mkRequest request =
+        headerList = request ^. reqHeaders
+
+renderHeader :: HeaderArg ElmDefinition -> Doc ann
+renderHeader headerInfo = "header" <+> "\"" <> headerName <> "\"" <+> parens ("withDefault" <+> "\"\"" <+> "headers." <> headerName)
```
This seems wrong. The semantics of `Maybe` is that the header **might not be present**, not that it might be present with an empty string for its value.

MaggieVL replied:
done

Commit:
Fix headerType name shadowing

Commit:
Use headers variable

Commit:
Add new endpoints to test headers

Commit:
Add spaces between headers

Commit:
Remove dashes from combinator names

The names are used when generating the function for its name and for the
name of the argument and they become invalid if dashed.

Commit:
Add new headers to corsPolicy

Commit:
Make type of header values String

We cannot toString each type in the elm query - e.g. Maybe a, so we
make the value a string before calling the function. We need to
stringify the values because headers have string values.

Commit:
Add commas to record type

Commit:
Add SortBy to elm-street types

So that the type, it's decoders and encoders can be generated

Commit:
Make Servant decode SortBy from urls

By making SortBy an instance of FromHttpApiData

Commit:
Make requests to the new endpoints and rename them

We rename everything so that it's clearer what we test with this request
(headers, body, query parameters, custom types, lists, etc).

Commit:
Add newly generated requests

Commit:
Move SorBy instancing next to its definition

Commit:
Describe why header values are passed as strings

googleson78 commented on examples/backend/src/ExampleAPI.hs:
```diff
@@ -1,10 +1,12 @@
+{-# LANGUAGE BangPatterns #-}
```
Where do you need this? I think you can remove it.

MaggieVL replied:
I used it when using the trace function in a let binding so that it is strictly evaluated, but it was only for testing, so I don't need it now.

MaggieVL replied:
done

googleson78 commented on examples/backend/src/ExampleAPI.hs:
```diff
@@ -39,8 +52,27 @@ data User = User
   deriving (Elm, ToJSON, FromJSON) via ElmStreet User
 
+data SortBy = Age | Name
+  deriving (Show, Generic, Elm)
+
+instance ToJSON SortBy
```
Use the `DerivingVia` style for `Elm` and `ToJSON`

MaggieVL replied:
done

googleson78 commented on examples/backend/src/ExampleAPI.hs:
```diff
@@ -57,11 +89,34 @@ albert = User {name = "Albert", age = 18}
-  return users
-    :<|> return albert
+  simpleRequests
     :<|> return
```
You should also factor this one, for consistency with the other handlers (and clarity too).

MaggieVL replied:
done

googleson78 commented on examples/backend/src/ExampleAPI.hs:
```diff
@@ -19,18 +21,29 @@ module ExampleAPI
+type Headers =
+  "basic" :> Header "someHeader" Text :> Post '[JSON] Text
+    :<|> "multiple" :> Header "someHeader1" Text :> Header' '[Required] "someHeader2" Int :> Post '[JSON] Text
+    :<|> "customType" :> Header "sortBy" SortBy :> Post '[JSON] SortBy
```
How about returning `Maybe SortBy` here, so that you can easily see if the header was read correctly?

MaggieVL replied:
as discussed yesterday, there is a Lenient header option that can be used in this case, so we don't need to pass Maybe values to headers

MaggieVL replied:
Do I need to test it with another type or still test it with a Maybe value?

googleson78 commented on src/ServantElm.hs:
```diff
@@ -132,10 +143,11 @@ endpointInfoToElmQuery requestInfo =
     bodyArg = case requestInfo ^. reqBody of
       Nothing -> []
       Just _ -> [bodyValue]
+    headerArgs = [headersValue | not (null $ requestInfo ^. reqHeaders)]
```
`not $ null $ ...`

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -205,6 +217,25 @@ mkTypeSignature request =
+
+    headersRecordType :: Maybe (Doc ann)
+    headersRecordType =
+      if null requestHeaders
+        then Nothing
+        else
+          Just $
+            braces $
+              concatWith
+                (surround (comma <> space))
+                (map headerToRecordField requestHeaders)
```
You can use `nonEmpty` instead here:
```haskell
nonEmpty requestHeaders <&> \requestHeaders' ->
  braces $
    concatWith
      (surround (comma <> space))
      (fmap headerToRecordField requestHeaders)
```

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -250,6 +281,24 @@ mkRequest request =
+      if null headerList
+        then "[]"
+        else
+          brackets $
+            concatWith
+              (surround (comma <> space))
```
You're using `brackets . concatWith (surround (comma <> space))` in two places, is there some nice name we can give to this function and factor it out?

MaggieVL replied:
done

Commit:
Remove BangPatterns extension

It was only for testing, not needed now.

Commit:
Use the deriving via syntax for SortBy

googleson78 commented on examples/frontend/src/Core/Generated/ElmQueries.elm:
```diff
@@ -54,16 +59,49 @@ postSignup urlBase toMsg bodyValue =
+        , tracker = Nothing
+        }
+
+postHeadersCustomType : String -> (Result Http.Error (SortBy) -> msg) -> {sortBy : String} -> Cmd msg
```
Since this header is optional in the api, the field should be `Maybe String` here.

MaggieVL replied:
done

Commit:
Handle optional headers server-side

When the header is optional, we need to pass maybe values, so that we
have the option of Nothing.

Commit:
Pass maybe values to request generating fns

Commit:
Add generated requests with both types of headers

Commit:
Use NonEmpty for the check if list is null

Commit:
Replace brackets with $

Commit:
Factor body request endpoints

For consistency with the other endpoints.

Commit:
Factor out a comma inserting function

googleson78 commented on README.md:
```diff
@@ -1,2 +1,7 @@
 # servant-elm-street
 Generate elm requests, with types, and JSON encoders, and decoders derived by elm-street, given a servant API.
+
+## Encoding headers
```
Document the assumption that a a user would never want to use `Maybe` as a header type, because the same effect could be achieved with the `Lenient` modifier.

Commit:
Document header Maybe values
